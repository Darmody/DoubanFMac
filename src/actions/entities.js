// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const save = (data: Object): Action => ({
  type: types.ENTITIES_SAVE,
  payload: data,
})

export const update = (data: Object): Action => ({
  type: types.ENTITIES_UPDATE,
  payload: data,
})
