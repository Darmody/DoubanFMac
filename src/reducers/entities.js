// @flow
import R from 'ramda'
import Immutable from 'seamless-immutable'
import * as types from 'constants/types/ActionTypes'
import type { Reducer } from 'constants/types/Redux'

const initialState = Immutable({
})

const reducer: Reducer = (state = initialState, { type, payload = {} }) => {
  switch (type) {
    case types.ENTITIES_SAVE: {
      let finalState = state
      R.forEachObjIndexed((data, model) =>
        R.forEachObjIndexed((value, id) => {
          finalState = Immutable.setIn(finalState, [model, id], value)
        })(data || {})
      )(payload)
      return finalState
    }
    case types.ENTITIES_UPDATE: {
      let finalState = state
      const updateField = ({ field, value }) => {
        finalState = Immutable.setIn(state, field, value)
      }

      R.forEach(updateField, payload)

      return finalState
    }
    default: return state
  }
}

export default reducer
