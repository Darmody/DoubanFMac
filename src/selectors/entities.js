// @flow
import { schema as Schema, denormalize } from 'normalizr'
import type { StateShape } from 'constants/types/Redux'

const isModelExists = (entities, model) => entities[model]

export const selectEntities = (state: StateShape, schema: Schema, ids: [] = []) => {
  const model = schema[0].key

  if (isModelExists(state.entities, model)) {
    return denormalize({ [model]: ids }, { [model]: schema }, state.entities)[model]
  }

  return []
}

export const selectEntity = (state: StateShape, schema: Schema, id: number = 0) => {
  const model = schema.key

  if (isModelExists(state.entities, model)) {
    return denormalize({ [model]: id }, { [model]: schema }, state.entities)[model]
  }

  return {}
}
