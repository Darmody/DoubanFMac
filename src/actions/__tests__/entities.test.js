import * as actions from '../entities'

describe('action::entities', () => {
  it('创建一个保存 entities action', () => {
    expect(actions.save({
      user: { 1: { id: 1 } }
    })).toMatchSnapshot()
  })

  it('创建一个更新 entity 字段 action', () => {
    expect(actions.update([{
      field: ['users', '1', 'id'], value: 2,
    }])).toMatchSnapshot()
  })
})
