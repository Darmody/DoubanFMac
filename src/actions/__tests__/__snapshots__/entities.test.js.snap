// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action::entities 创建一个保存 entities action 1`] = `
Object {
  "payload": Object {
    "user": Object {
      "1": Object {
        "id": 1,
      },
    },
  },
  "type": "ENTITIES/SAVE",
}
`;

exports[`action::entities 创建一个更新 entity 字段 action 1`] = `
Object {
  "payload": Array [
    Object {
      "field": Array [
        "users",
        "1",
        "id",
      ],
      "value": 2,
    },
  ],
  "type": "ENTITIES/UPDATE",
}
`;
